File: python/src/expression_parser/expression.py
# This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
# Copyright (c) 2025 Ian Thomas

import inspect
from abc import abstractmethod
from typing import Any, Dict, List, Optional, Union

STRING_FORMAT_SINGLEQUOTE = 0
STRING_FORMAT_ESCAPED_SINGLEQUOTE = 1
STRING_FORMAT_DOUBLEQUOTE = 2
STRING_FORMAT_ESCAPED_DOUBLEQUOTE = 3

class ExpressionNode:
    @abstractmethod
    def __init__(self, name: str) -> None:
        self._name = name

    @abstractmethod
    def evaluate(self, context: Dict[str, Any], dump_eval: Optional[List[str]] = None) -> Any:
        raise NotImplementedError()

    @abstractmethod
    def dump_structure(self, indent: int = 0) -> str:
        raise NotImplementedError()
    
    @abstractmethod
    def write(self, string_format = STRING_FORMAT_SINGLEQUOTE) -> str:
        raise NotImplementedError()
    

class BinaryOp(ExpressionNode):
    @abstractmethod
    def __init__(self, name: str, left: ExpressionNode, op: str, right: ExpressionNode) -> None:
        super().__init__(name)
        self._left = left
        self._op = op
        self._right = right
    
    def evaluate(self, context: Dict[str, Any], dump_eval: Optional[List[str]] = None) -> Any:
        left_val = self._left.evaluate(context, dump_eval)
        right_val = self._right.evaluate(context, dump_eval)
        result: Any = self._do_eval(left_val, right_val)
        
        if dump_eval is not None:
            dump_eval.append(f"Evaluated: {left_val} {self._op} {right_val} = {result}")

        return result
    
    @abstractmethod
    def _do_eval(self, left_val: Any, right_val: Any) -> Any:
        raise NotImplementedError()

    def dump_structure(self, indent: int = 0) -> str:
        out = ("  " * indent + f"{self._name}") + "\n"
        out += self._left.dump_structure(indent + 1)
        out += self._right.dump_structure(indent + 1)
        return out
    
    def write(self, string_format = STRING_FORMAT_SINGLEQUOTE) -> str:
        return f"{self._left.write(string_format)} {self._op} {self._right.write(string_format)}"
    

class OpAnd(BinaryOp):
    def __init__(self, left: ExpressionNode, right: ExpressionNode) -> None:
        super().__init__("And", left, "and", right)

    def _do_eval(self, left_val: Any, right_val: Any) -> Any: 
        return _make_bool(left_val) and _make_bool(right_val)
    

class OpOr(BinaryOp):
    def __init__(self, left: ExpressionNode, right: ExpressionNode) -> None:
        super().__init__("Or", left, "or", right)

    def _do_eval(self, left_val: Any, right_val: Any) -> Any: 
        return _make_bool(left_val) or _make_bool(right_val)


class OpEquals(BinaryOp):
    def __init__(self, left: ExpressionNode, right: ExpressionNode) -> None:
        super().__init__("Equals", left, "==", right)

    def _do_eval(self, left_val: Any, right_val: Any) -> Any: 
        right_val = _make_type_match(left_val, right_val)
        return left_val == right_val


class OpNotEquals(BinaryOp):
    def __init__(self, left: ExpressionNode, right: ExpressionNode) -> None:
        super().__init__("NotEquals", left, "!=", right)

    def _do_eval(self, left_val: Any, right_val: Any) -> Any: 
        right_val = _make_type_match(left_val, right_val)
        return left_val != right_val


class OpPlus(BinaryOp):
    def __init__(self, left: ExpressionNode, right: ExpressionNode) -> None:
        super().__init__("Plus", left, "+", right)

    def _do_eval(self, left_val: Any, right_val: Any) -> Any: 
        return _make_numeric(left_val) + _make_numeric(right_val)


class OpMinus(BinaryOp):
    def __init__(self, left: ExpressionNode, right: ExpressionNode) -> None:
        super().__init__("Minus", left, "-", right)

    def _do_eval(self, left_val: Any, right_val: Any) -> Any: 
        return _make_numeric(left_val) - _make_numeric(right_val)


class OpMultiply(BinaryOp):
    def __init__(self, left: ExpressionNode, right: ExpressionNode) -> None:
        super().__init__("Multiply", left, "*", right)

    def _do_eval(self, left_val: Any, right_val: Any) -> Any: 
        return _make_numeric(left_val) * _make_numeric(right_val)
    

class OpDivide(BinaryOp):
    def __init__(self, left: ExpressionNode, right: ExpressionNode) -> None:
        super().__init__("Divide", left, "/", right)

    def _do_eval(self, left_val: Any, right_val: Any) -> Any: 
        right_val = _make_numeric(right_val)
        if right_val == 0:
            raise ZeroDivisionError("Division by zero.")
        return _make_numeric(left_val) / right_val


class OpGreaterThan(BinaryOp):
    def __init__(self, left: ExpressionNode, right: ExpressionNode) -> None:
        super().__init__("GreaterThan", left, ">", right)

    def _do_eval(self, left_val: Any, right_val: Any) -> Any: 
        return _make_numeric(left_val) > _make_numeric(right_val)


class OpLessThan(BinaryOp):
    def __init__(self, left: ExpressionNode, right: ExpressionNode) -> None:
        super().__init__("LessThan", left, "<", right)

    def _do_eval(self, left_val: Any, right_val: Any) -> Any: 
        return _make_numeric(left_val) < _make_numeric(right_val)


class OpGreaterThanEquals(BinaryOp):
    def __init__(self, left: ExpressionNode, right: ExpressionNode) -> None:
        super().__init__("GreaterThanEquals", left, ">=", right)
    
    def _do_eval(self, left_val: Any, right_val: Any) -> Any: 
        return _make_numeric(left_val) >= _make_numeric(right_val)
    
    
class OpLessThanEquals(BinaryOp):
    def __init__(self, left: ExpressionNode, right: ExpressionNode) -> None:
        super().__init__("LessThanEquals", left, "<=", right)

    def _do_eval(self, left_val: Any, right_val: Any) -> Any: 
        return _make_numeric(left_val) <= _make_numeric(right_val)
    

class UnaryOp(ExpressionNode):
    @abstractmethod
    def __init__(self, name: str, op: str, operand: ExpressionNode) -> None:
        super().__init__(name)
        self._operand = operand
        self._op = op

    def evaluate(self, context: Dict[str, Any], dump_eval: Optional[List[str]] = None) -> Any:
        val = self._operand.evaluate(context, dump_eval)
        result: Any = self._do_eval(val)
        
        if dump_eval is not None:
            dump_eval.append(f"Evaluated: {self._op} {val} = {result}")
        
        return result
    
    @abstractmethod
    def _do_eval(self, val: Any) -> Any:
        raise NotImplementedError()

    def dump_structure(self, indent: int = 0) -> str:
        out = ("  " * indent + f"{self._name}") + "\n"
        out += self._operand.dump_structure(indent + 1)
        return out
    
    def write(self, string_format = STRING_FORMAT_SINGLEQUOTE) -> str:
        return self._op + self._operand.write(string_format)
    

class OpNegative(UnaryOp):
    def __init__(self, operand: ExpressionNode) -> None:
        super().__init__("Negative", "-", operand)

    def _do_eval(self, val: Any) -> Any:
        val = _make_numeric(val)
        return -val


class OpNot(UnaryOp):
    def __init__(self, operand: ExpressionNode) -> None:
        super().__init__("Not", "not", operand)

    def _do_eval(self, val: Any) -> Any:
        val = _make_bool(val)
        return not val


class LiteralBoolean(ExpressionNode):
    def __init__(self, value: bool) -> None:
        super().__init__("Boolean")
        self._value = value

    def evaluate(self, context: Dict[str, Any], dump_eval: Optional[List[str]] = None) -> bool:
        if dump_eval is not None:
            dump_eval.append(f"Boolean: {self._value}")
        return self._value

    def dump_structure(self, indent: int = 0) -> str:
        return ("  " * indent + f"Boolean({self._value})") + "\n"
    
    def write(self, string_format = STRING_FORMAT_SINGLEQUOTE) -> str:
        return "true" if self._value else "false"


class LiteralNumber(ExpressionNode):
    def __init__(self, value: str) -> None:
        super().__init__("Number")
        self._value = float(value) if '.' in value else int(value)

    def evaluate(self, context: Dict[str, Any], dump_eval: Optional[List[str]] = None) -> Union[int, float]:
        if dump_eval is not None:
            dump_eval.append(f"Number: {self._value}")
        return self._value

    def dump_structure(self, indent: int = 0) -> str:
        return ("  " * indent + f"Number({self._value})") + "\n"

    def write(self, string_format = STRING_FORMAT_SINGLEQUOTE) -> str:
        return str(self._value)
    

class LiteralString(ExpressionNode):
    def __init__(self, value: str) -> None:
        super().__init__("String")
        self._value = value

    def evaluate(self, context: Dict[str, Any], dump_eval: Optional[List[str]] = None) -> str:
        if dump_eval is not None:
            dump_eval.append(f"String: {self._value}")
        return self._value

    def dump_structure(self, indent: int = 0) -> str:
        return ("  " * indent + f"String({self._value})") + "\n"

    def write(self, string_format = STRING_FORMAT_SINGLEQUOTE) -> str:
        if string_format==STRING_FORMAT_SINGLEQUOTE:
            return f"'{self._value}'"
        if string_format==STRING_FORMAT_ESCAPED_SINGLEQUOTE:
            return f"\\'{self._value}\\'"
        if string_format==STRING_FORMAT_ESCAPED_DOUBLEQUOTE:
            return f"\\\"{self._value}\\\""
        return f"\"{self._value}\""
    
    
class Variable(ExpressionNode):
    def __init__(self, name: str) -> None:
        super().__init__("Variable")
        self._name = name

    def evaluate(self, context: Dict[str, Any], dump_eval: Optional[List[str]] = None) -> Any:
        value = context.get(self._name)
        if value is None:
            raise RuntimeError(f"Variable '{self._name}' not found in context.")
        if not isinstance(value, (int, float, bool, str)):
            raise TypeError(f"Variable '{self._name}' must return bool, string, or numeric.")
        
        if dump_eval is not None:
            dump_eval.append(f"Fetching variable: {self._name} -> {value}")
        return value

    def dump_structure(self, indent: int = 0) -> str:
        return ("  " * indent + f"Variable({self._name})") + "\n"
    
    def write(self, string_format = STRING_FORMAT_SINGLEQUOTE) -> str:
        return self._name
    

class FunctionCall(ExpressionNode):
    def __init__(self, func_name: str, args: Optional[List[ExpressionNode]] = None) -> None:
        super().__init__("FunctionCall")
        if args is None:
            args = []
        self._func_name = func_name
        self._args = args

    def evaluate(self, context: Dict[str, Any], dump_eval: Optional[List[str]] = None) -> Any:
        func = context.get(self._func_name)
        if func is None:
            raise RuntimeError(f"Function '{self._func_name}' not found in context.")
        
        arg_values: List[Any] = [arg.evaluate(context, dump_eval) for arg in self._args]

        # Get the function signature and check if it accepts the provided arguments.
        sig = inspect.signature(func)
        try:
            sig.bind(*arg_values)
        except TypeError as e:
            raise RuntimeError(f"Function '{self._func_name}' does not support the provided arguments {arg_values}.")

        result = func(*arg_values)

        if not isinstance(result, (int, float, bool, str)):
            raise TypeError(f"Function '{self._func_name}' must return bool, string, or numeric.")
        
        if dump_eval is not None:
            dump_eval.append(f"Calling function: {self._func_name}({arg_values}) = {result}")

        return result

    def dump_structure(self, indent: int = 0) -> str:
        out = ("  " * indent + f"FunctionCall({self._func_name})") + "\n"
        for arg in self._args:
            out += arg.dump_structure(indent + 1)
        return out
    
    def write(self, string_format = STRING_FORMAT_SINGLEQUOTE) -> str:
        out = self._func_name+"("
        written_args = []
        for arg in self._args:
            written_args.append(arg.write(string_format))
        out += ", ".join(written_args)
        out += ")"
        return out
    

def _make_bool(val: Any) -> bool:
    if isinstance(val, bool):
        return val
    if isinstance(val, (int, float)):
        return val != 0
    if isinstance(val, str):
        return val.lower() == "true" or val == "1"
    raise TypeError(f"Type mismatch: Expecting bool, but got '{val}'")


def _make_str(val: Any) -> str:
    if isinstance(val, str):
        return val
    if isinstance(val, bool):
        return "true" if val else "false"
    if isinstance(val, (int, float)):
        return str(val)
    raise TypeError(f"Type mismatch: Expecting string but got '{val}'")


def _make_numeric(val: Any) -> Union[int, float]:
    if isinstance(val, bool):
        return 1 if val else 0
    if isinstance(val, (float, int)):
        return val
    if isinstance(val, str):
        try:
            return int(val)
        except ValueError:
            try:
                return float(val)
            except ValueError:
                pass
    raise TypeError(f"Type mismatch: Expecting number but got '{val}'")


def _make_type_match(left_val: Any, right_val: Any) -> Any:
    if isinstance(left_val, bool):
        return _make_bool(right_val)
    if isinstance(left_val, (int, float)):
        return _make_numeric(right_val)
    if isinstance(left_val, str):
        return _make_str(right_val)
    raise TypeError(f"Type mismatch: unrecognised type for '{left_val}'")

File: javascript/src/expression.js
/*
 * This file is part of an MIT-licensed project: see LICENSE file or README.md for details.
 * Copyright (c) 2025 Ian Thomas
 */

export const StringFormat = Object.freeze({
  SINGLEQUOTE: 0,
  ESCAPED_SINGLEQUOTE: 1,
  DOUBLEQUOTE: 2,
  ESCAPED_DOUBLEQUOTE: 3,
});

let _STRING_FORMAT = StringFormat.SINGLEQUOTE;

export function setStringFormat(newFormat) {
  _STRING_FORMAT = newFormat;
}

export class ExpressionNode {
  constructor(name, precedence) {
    if (new.target === ExpressionNode) {
      throw new TypeError("Cannot instantiate abstract class ExpressionNode directly");
    }
    this._name = name;
    this._precedence = precedence;
  }

  evaluate(context, dump_eval) {
    throw new Error("Abstract method 'evaluate' not implemented");
  }

  dump_structure(indent = 0) {
    throw new Error("Abstract method 'dump_structure' not implemented");
  }

  write() {
    throw new Error("Abstract method 'write' not implemented");
  }
}

export class BinaryOp extends ExpressionNode {
  constructor(name, left, op, right, precedence) {
    super(name, precedence);
    this._left = left;
    this._op = op;
    this._right = right;
  }

  evaluate(context, dump_eval) {
    const left_val = this._left.evaluate(context, dump_eval);
    const right_val = this._right.evaluate(context, dump_eval);
    const result = this._do_eval(left_val, right_val);

    if (dump_eval) {
      dump_eval.push(`Evaluated: ${_format_value(left_val)} ${this._op} ${_format_value(right_val)} = ${_format_value(result)}`);
    }
    return result;
  }

  _do_eval(left_val, right_val) {
    throw new Error("Abstract method '_do_eval' not implemented");
  }

  dump_structure(indent = 0) {
    let out = "  ".repeat(indent) + `${this._name}\n`;
    out += this._left.dump_structure(indent + 1);
    out += this._right.dump_structure(indent + 1);
    return out;
  }

  write() {
    let left_str = this._left.write();
    let right_str = this._right.write();

    if (this._left._precedence < this._precedence) {
      left_str = `(${left_str})`;
    }
    if (this._right._precedence < this._precedence) {
      right_str = `(${right_str})`;
    }
    return `${left_str} ${this._op} ${right_str}`;
  }
}

export class OpOr extends BinaryOp {
  constructor(left, right) {
    super("Or", left, "or", right, 40);
  }
  _do_eval(left_val, right_val) {
    return _make_bool(left_val) || _make_bool(right_val);
  }
}

export class OpAnd extends BinaryOp {
  constructor(left, right) {
    super("And", left, "and", right, 50);
  }
  _do_eval(left_val, right_val) {
    return _make_bool(left_val) && _make_bool(right_val);
  }
}

export class OpEquals extends BinaryOp {
  constructor(left, right) {
    super("Equals", left, "==", right, 60);
  }
  _do_eval(left_val, right_val) {
    right_val = _make_type_match(left_val, right_val);
    return left_val === right_val;
  }
}

export class OpNotEquals extends BinaryOp {
  constructor(left, right) {
    super("NotEquals", left, "!=", right, 60);
  }
  _do_eval(left_val, right_val) {
    right_val = _make_type_match(left_val, right_val);
    return left_val !== right_val;
  }
}

export class OpPlus extends BinaryOp {
  constructor(left, right) {
    super("Plus", left, "+", right, 70);
  }
  _do_eval(left_val, right_val) {
    return _make_numeric(left_val) + _make_numeric(right_val);
  }
}

export class OpMinus extends BinaryOp {
  constructor(left, right) {
    super("Minus", left, "-", right, 70);
  }
  _do_eval(left_val, right_val) {
    return _make_numeric(left_val) - _make_numeric(right_val);
  }
}

export class OpDivide extends BinaryOp {
  constructor(left, right) {
    super("Divide", left, "/", right, 85);
  }
  _do_eval(left_val, right_val) {
    right_val = _make_numeric(right_val);
    if (right_val === 0) {
      throw new Error("Division by zero.");
    }
    return _make_numeric(left_val) / right_val;
  }
}

export class OpMultiply extends BinaryOp {
  constructor(left, right) {
    super("Multiply", left, "*", right, 80);
  }
  _do_eval(left_val, right_val) {
    return _make_numeric(left_val) * _make_numeric(right_val);
  }
}

export class OpGreaterThan extends BinaryOp {
  constructor(left, right) {
    super("GreaterThan", left, ">", right, 60);
  }
  _do_eval(left_val, right_val) {
    return _make_numeric(left_val) > _make_numeric(right_val);
  }
}

export class OpLessThan extends BinaryOp {
  constructor(left, right) {
    super("LessThan", left, "<", right, 60);
  }
  _do_eval(left_val, right_val) {
    return _make_numeric(left_val) < _make_numeric(right_val);
  }
}

export class OpGreaterThanEquals extends BinaryOp {
  constructor(left, right) {
    super("GreaterThanEquals", left, ">=", right, 60);
  }
  _do_eval(left_val, right_val) {
    return _make_numeric(left_val) >= _make_numeric(right_val);
  }
}

export class OpLessThanEquals extends BinaryOp {
  constructor(left, right) {
    super("LessThanEquals", left, "<=", right, 60);
  }
  _do_eval(left_val, right_val) {
    return _make_numeric(left_val) <= _make_numeric(right_val);
  }
}

export class UnaryOp extends ExpressionNode {
  constructor(name, op, operand, precedence) {
    super(name, precedence);
    this._operand = operand;
    this._op = op;
  }

  evaluate(context, dump_eval) {
    const val = this._operand.evaluate(context, dump_eval);
    const result = this._do_eval(val);
    if (dump_eval) {
      dump_eval.push(`Evaluated: ${this._op} ${_format_value(val)} = ${_format_value(result)}`);
    }
    return result;
  }

  _do_eval(val) {
    throw new Error("Abstract method '_do_eval' not implemented");
  }

  dump_structure(indent = 0) {
    let out = "  ".repeat(indent) + `${this._name}\n`;
    out += this._operand.dump_structure(indent + 1);
    return out;
  }

  write() {
    let operand_str = this._operand.write();
    if (this._operand._precedence < this._precedence) {
      operand_str = `(${operand_str})`;
    }
    return `${this._op} ${operand_str}`;
  }
}

export class OpNegative extends UnaryOp {
  constructor(operand) {
    super("Negative", "-", operand, 90);
  }
  _do_eval(val) {
    return -_make_numeric(val);
  }
}

export class OpNot extends UnaryOp {
  constructor(operand) {
    super("Not", "not", operand, 90);
  }
  _do_eval(val) {
    return !_make_bool(val);
  }
}

export class LiteralBoolean extends ExpressionNode {
  constructor(value) {
    super("Boolean", 100);
    this._value = value;
  }
  evaluate(context, dump_eval) {
    if (dump_eval) {
      dump_eval.push(`Boolean: ${_format_value(this._value)}`);
    }
    return this._value;
  }
  dump_structure(indent = 0) {
    return "  ".repeat(indent) + `Boolean(${_format_value(this._value)})\n`;
  }
  write() {
    return _format_value(this._value);
  }
}

export class LiteralNumber extends ExpressionNode {
  constructor(value) {
    super("Number", 100);
    // Assuming value is provided as a string
    this._value = parseFloat(value);
  }
  evaluate(context, dump_eval) {
    if (dump_eval) {
      dump_eval.push(`Number: ${this._value}`);
    }
    return this._value;
  }
  dump_structure(indent = 0) {
    return "  ".repeat(indent) + `Number(${this._value})\n`;
  }
  write() {
    return `${this._value}`;
  }
}

export class LiteralString extends ExpressionNode {
  constructor(value) {
    super("String", 100);
    this._value = value;
  }
  evaluate(context, dump_eval) {
    if (dump_eval) {
      dump_eval.push(`String: ${_format_string(this._value)}`);
    }
    return this._value;
  }
  dump_structure(indent = 0) {
    return "  ".repeat(indent) + `String(${_format_string(this._value)})\n`;
  }
  write() {
    return _format_string(this._value);
  }
}

export class Variable extends ExpressionNode {
  constructor(name) {
    super("Variable", 100);
    this._name = name;
  }
  evaluate(context, dump_eval) {
    const value = context[this._name];
    if (value === undefined) {
      throw new Error(`Variable '${this._name}' not found in context.`);
    }
    if (typeof value !== "number" && typeof value !== "boolean" && typeof value !== "string") {
      throw new TypeError(`Variable '${this._name}' must return bool, string, or numeric.`);
    }
    if (dump_eval) {
      dump_eval.push(`Fetching variable: ${this._name} -> ${_format_value(value)}`);
    }
    return value;
  }
  dump_structure(indent = 0) {
    return "  ".repeat(indent) + `Variable(${this._name})\n`;
  }
  write() {
    return this._name;
  }
}

export class FunctionCall extends ExpressionNode {
  constructor(func_name, args = []) {
    super("FunctionCall", 100);
    this._func_name = func_name;
    this._args = args;
  }
  evaluate(context, dump_eval) {
    const func = context[this._func_name];
    if (func === undefined) {
      throw new Error(`Function '${this._func_name}' not found in context.`);
    }
    const arg_values = this._args.map(arg => arg.evaluate(context, dump_eval));

    // Check function arity using func.length
    if (arg_values.length !== func.length) {
      const formattedArgs = arg_values.map(val => _format_value(val)).join(", ");
      throw new Error(`Function '${this._func_name}' does not support the provided arguments (${formattedArgs}).`);
    }

    const result = func(...arg_values);

    if (typeof result !== "number" && typeof result !== "boolean" && typeof result !== "string") {
      throw new TypeError(`Function '${this._func_name}' must return bool, string, or numeric.`);
    }
    if (dump_eval) {
      const formattedArgs = arg_values.map(val => _format_value(val)).join(", ");
      dump_eval.push(`Called function: ${this._func_name}(${formattedArgs}) = ${_format_value(result)}`);
    }
    return result;
  }
  dump_structure(indent = 0) {
    let out = "  ".repeat(indent) + `FunctionCall(${this._func_name})\n`;
    for (const arg of this._args) {
      out += arg.dump_structure(indent + 1);
    }
    return out;
  }
  write() {
    const written_args = this._args.map(arg => arg.write());
    return `${this._func_name}(${written_args.join(", ")})`;
  }
}

function _make_bool(val) {
  if (typeof val === "boolean") {
    return val;
  }
  if (typeof val === "number") {
    return val !== 0;
  }
  if (typeof val === "string") {
    return val.toLowerCase() === "true" || val === "1";
  }
  throw new TypeError(`Type mismatch: Expecting bool, but got '${val}'`);
}

function _make_str(val) {
  if (typeof val === "string") {
    return val;
  }
  if (typeof val === "boolean") {
    return val ? "true" : "false";
  }
  if (typeof val === "number") {
    return val.toString();
  }
  throw new TypeError(`Type mismatch: Expecting string but got '${val}'`);
}

function _make_numeric(val) {
  if (typeof val === "boolean") {
    return val ? 1 : 0;
  }
  if (typeof val === "number") {
    return val;
  }
  if (typeof val === "string") {
    const numericRegex = /^-?\d+(\.\d+)?$/;
    if (numericRegex.test(val)) {
      return parseFloat(val);
    }
  }
  throw new TypeError(`Type mismatch: Expecting number but got '${val}'`);
}

function _make_type_match(left_val, right_val) {
  if (typeof left_val === "boolean") {
    return _make_bool(right_val);
  }
  if (typeof left_val === "number") {
    return _make_numeric(right_val);
  }
  if (typeof left_val === "string") {
    return _make_str(right_val);
  }
  throw new TypeError(`Type mismatch: unrecognised type for '${left_val}'`);
}

function _format_string(val) {
  switch (_STRING_FORMAT) {
    case StringFormat.SINGLEQUOTE:
      return `'${val}'`;
    case StringFormat.ESCAPED_SINGLEQUOTE:
      return `\\'${val}\\'`;
    case StringFormat.ESCAPED_DOUBLEQUOTE:
      return `\\"${val}\\"`;
    case StringFormat.DOUBLEQUOTE:
    default:
      return `"${val}"`;
  }
}

function _format_value(val) {
  if (typeof val === "string") {
    return _format_string(val);
  }
  return val.toString();
}